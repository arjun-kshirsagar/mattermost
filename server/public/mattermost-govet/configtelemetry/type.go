// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

package configtelemetry

import (
	"go/ast"
	"go/token"
	"strings"
)

var ignoreComments = []string{
	"telemetry: none",
	"Deprecated: do not use",
}

// typeFieldMap will generate map of its all possible selectors and adds them as a key. The value
// is the position of the declaration in the type spec.
func typeFieldMap(file *ast.File, typeName string) (map[string]token.Pos, error) {

	var fieldsMap map[string]token.Pos

	ast.Inspect(file, func(n ast.Node) bool {
		decl, ok := n.(*ast.GenDecl)
		if !ok || decl.Tok != token.TYPE || len(decl.Specs) < 1 {
			return true
		}

		spec, ok := decl.Specs[0].(*ast.TypeSpec)
		if !ok || spec.Name.Name != typeName {
			return true
		}

		typ, ok := spec.Type.(*ast.StructType)
		if !ok {
			return true
		}

		fieldsMap = fields(typ)

		return false
	})

	return fieldsMap, nil
}

func fields(str *ast.StructType) map[string]token.Pos {
	structFields := make(map[string]token.Pos)

	for _, field := range str.Fields.List {
		if field.Comment != nil && ignored(field.Comment) {
			continue
		}

		expr := field.Type
		if ptr, ok := field.Type.(*ast.StarExpr); ok {
			expr = ptr.X
		}

		if len(field.Names) != 1 {
			panic("unhandled struct field definition")
		}

		key := field.Names[0].Name

		v, ok := expr.(*ast.Ident)
		if !ok || v.Obj == nil || v.Obj.Decl == nil {
			structFields[key] = field.Pos()
			continue
		}

		typeDecl, ok := v.Obj.Decl.(*ast.TypeSpec)
		if !ok {
			continue
		}

		typ, ok := typeDecl.Type.(*ast.StructType)
		if !ok {
			continue
		}

		for k, v := range fields(typ) {
			structFields[key+"."+k] = v
		}
	}

	return structFields
}

func ignored(comments *ast.CommentGroup) bool {
	for _, comment := range comments.List {
		for _, ignore := range ignoreComments {
			if strings.Contains(comment.Text, ignore) {
				return true
			}
		}
	}
	return false
}
