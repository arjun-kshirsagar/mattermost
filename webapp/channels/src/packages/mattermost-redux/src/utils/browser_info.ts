// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export function getBrowserInfo() {
    const userAgent = window.navigator.userAgent;

    let browser = 'Unknown';
    let browserVersion = 'Unknown';

    // Simple browser detection
    if (userAgent.includes('Firefox')) {
        browser = 'Firefox';
    } else if (userAgent.includes('Chrome')) {
        browser = 'Chrome';
    } else if (userAgent.includes('Safari')) {
        browser = 'Safari';
    } else if (userAgent.includes('Edge')) {
        browser = 'Edge';
    }

    // Get browser version
    const match = userAgent.match(/(firefox|chrome|safari|edge(?=\/))\/?\s*(\d+)/i);
    if (match) {
        browserVersion = match[2];
    }

    return {browser, browserVersion};
}

export function getPlatformInfo() {
    // Casting to undefined in case it is deprecated in any browser
    const platform = window.navigator.platform as string | undefined;

    let platformName = 'Unknown';
    if (platform?.toLowerCase().includes('win')) {
        platformName = 'Windows';
    } else if (platform?.toLowerCase().includes('mac')) {
        platformName = 'MacOS';
    } else if (platform?.toLowerCase().includes('linux')) {
        platformName = 'Linux';
    }

    return platformName;
}
