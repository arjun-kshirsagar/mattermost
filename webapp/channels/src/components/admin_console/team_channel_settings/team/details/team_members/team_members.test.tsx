// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

// No direct testing library imports needed
import React from 'react';

import type {Team, TeamMembership} from '@mattermost/types/teams';
import type {UserProfile} from '@mattermost/types/users';

import {renderWithContext} from 'tests/react_testing_utils';
import {TestHelper} from 'utils/test_helper';

import TeamMembers from './team_members';

// Mock components to simplify testing
jest.mock('components/admin_console/user_grid/user_grid', () => {
    return function MockUserGrid() {
        return <div data-testid='user-grid'/>;
    };
});

jest.mock('components/widgets/admin_console/admin_panel', () => {
    return function MockAdminPanel(props: any) {
        return (
            <div data-testid='admin-panel'>
                <div className='header'>{typeof props.title === 'object' ? props.title.defaultMessage : props.title}</div>
                <div className='body'>{props.children}</div>
            </div>
        );
    };
});

jest.mock('components/toggle_modal_button', () => {
    return function MockToggleModalButton(props: any) {
        return <button data-testid='toggle-modal-button'>{props.children}</button>;
    };
});

describe('admin_console/team_channel_settings/team/TeamMembers', () => {
    const user1: UserProfile = TestHelper.getUserMock({id: 'user-1'});
    const membership1: TeamMembership = TestHelper.getTeamMembershipMock({user_id: 'user-1'});
    const user2: UserProfile = TestHelper.getUserMock({id: 'user-2'});
    const membership2: TeamMembership = TestHelper.getTeamMembershipMock({user_id: 'user-2'});
    const user3: UserProfile = TestHelper.getUserMock({id: 'user-3'});
    const membership3: TeamMembership = TestHelper.getTeamMembershipMock({user_id: 'user-3'});
    const team: Team = TestHelper.getTeamMock({id: 'team-1'});

    const baseProps = {
        filters: {},
        teamId: 'team-1',
        team,
        users: [user1, user2, user3],
        usersToRemove: {},
        usersToAdd: {},
        teamMembers: {
            [user1.id]: membership1,
            [user2.id]: membership2,
            [user3.id]: membership3,
        },
        enableGuestAccounts: true,

        totalCount: 3,
        loading: false,
        searchTerm: '',
        onAddCallback: jest.fn(),
        onRemoveCallback: jest.fn(),
        updateRole: jest.fn(),

        actions: {
            getTeamStats: jest.fn().mockResolvedValue({}),
            loadProfilesAndReloadTeamMembers: jest.fn().mockResolvedValue({}),
            searchProfilesAndTeamMembers: jest.fn().mockResolvedValue({}),
            getFilteredUsersStats: jest.fn().mockResolvedValue({}),
            setUserGridSearch: jest.fn().mockResolvedValue({}),
            setUserGridFilters: jest.fn().mockResolvedValue({}),
        },
    };

    test('should render properly with users', () => {
        const {container, getByTestId} = renderWithContext(
            <TeamMembers {...baseProps}/>,
        );

        expect(getByTestId('admin-panel')).toBeInTheDocument();
        expect(getByTestId('user-grid')).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    test('should render properly when loading with no users', () => {
        const props = {
            ...baseProps,
            users: [],
            teamMembers: {},
            totalCount: 0,
            loading: true,
        };

        const {container, getByTestId} = renderWithContext(
            <TeamMembers {...props}/>,
        );

        expect(getByTestId('admin-panel')).toBeInTheDocument();
        expect(getByTestId('user-grid')).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });
});
