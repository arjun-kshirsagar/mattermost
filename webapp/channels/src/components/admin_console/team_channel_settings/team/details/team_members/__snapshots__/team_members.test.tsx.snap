// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`admin_console/team_channel_settings/team/TeamMembers should render properly when loading with no users 1`] = `
<div>
  <div
    data-testid="admin-panel"
  >
    <div
      class="header"
    >
      Members
    </div>
    <div
      class="body"
    >
      <div
        data-testid="user-grid"
      />
    </div>
  </div>
</div>
`;

exports[`admin_console/team_channel_settings/team/TeamMembers should render properly with users 1`] = `
<div>
  <div
    data-testid="admin-panel"
  >
    <div
      class="header"
    >
      Members
    </div>
    <div
      class="body"
    >
      <div
        data-testid="user-grid"
      />
    </div>
  </div>
</div>
`;
